model/renderNativeWindow.cc

```diff
@@ -54,41 +63,69 @@ using namespace ecolab;
 namespace minsky
 {
 #ifdef USE_WIN32_SURFACE
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
-  {/* TODO */}
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &winInfo)
+  { /* TODO */
+  }
 #elif defined(MAC_OSX_TK)
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
-  {/* TODO */}
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &winInfo)
+  { /* TODO */
+  }
 #else
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &wi)
+  {
+    cairo::SurfacePtr childSurface(new cairo::Surface(cairo_xlib_surface_create(wi.getDisplay(), wi.getChildWindowId(), wi.wAttr.visual, wi.childWidth, wi.childHeight), wi.childWidth, wi.childHeight));
+    cairo_surface_set_device_offset(childSurface->surface(), -wi.wAttr.x, -wi.wAttr.y);
+    return childSurface;
+  }
+#endif
 
-  int throwOnXError(Display*, XErrorEvent* ev)
+  RenderNativeWindow::~RenderNativeWindow()
   {
-    char errorMessage[256];
-    XGetErrorText(ev->display, ev->error_code, errorMessage, sizeof(errorMessage));
-    throw runtime_error(errorMessage);
+    if (this->winInfo)
+    {
+      delete this->winInfo;
+    }
   }
-  
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
+
+  RenderNativeWindow::RenderNativeWindow()
   {
-    // ensure errors are thrown, rather than exit() being called
-    static bool errorHandlingSet=(XSetErrorHandler(throwOnXError), true);
-    XWindowAttributes wAttr;
-    auto display=XOpenDisplay(nullptr);
-    int err=XGetWindowAttributes(display, window, &wAttr);
-    if (err>1)
-      throw runtime_error("Invalid window: "+to_string(window));
-    cairo::SurfacePtr r(new cairo::Surface(cairo_xlib_surface_create(display,window,wAttr.visual,wAttr.width,wAttr.height),wAttr.width,wAttr.height));
-    cairo_surface_set_device_offset(r->surface(), -wAttr.x, -wAttr.y);
-    return r;
+    this->winInfo = new WindowInformation();
   }
-#endif
-  
 
-  void RenderNativeWindow::renderToNativeWindow(unsigned long window)
+  RenderNativeWindow &RenderNativeWindow::operator=(const RenderNativeWindow &a)
   {
-    auto tmp=nativeWindowSurface(window);
+    //TODO:: I expected this to be the "new" instance with uninitialized winInfo and a to have the previous winInfo.
+    // However, it seems to be the other way round in practice --- Janak
+    a.winInfo->initialize(*(this->winInfo));
```

Comment:
In C++, just because a pointer is const, doesn't mean the target of the pointer is const. It's a little counterintuitive, but you have to add extra qualifiers to the type to ensure the target is const,

Nevertheless, it is is bad form to go changing the RHS of an assignment statement - it would be totally surprising to most C++ programmers.